### DISCUSSION TITLE
Sequencial animations

### DISCUSSION THREAD

**Discussion by Mark2Mark** [coldtype/coldtype#75] [Q&A] [answered]
Hi there!

Totally loving Coldtype and its capabilities.

I understand that the documentation might have a little lack in one or two spots, I am also happy to help.

I wonder if there’s a way to construct sequencial animations. I dug through the examples and source code, to get some ideas. The only thing related, that I could find is the `state` array with dictionaries, which seems to must have numeral values as keys. Then `@animation(timeline=Loop(60, 10, len(states))` plus 
```py
phase = f.a.t.current_phase(f.i)
state = phase.calc_state(states)
```
stitches the states nicely together. This is awesome.

But what I wonder is, if there’s a way to create a kind of timeline with the states having different lengths. This could unlock creating rather complex timelines, somewhat similar to the midi examples, just not driven by a midi, but by a dictionary of some kind.

Maybe there’s somehting already, and I just didn’t see it.

Anyway, amazing tool, Coldtype is! Thanks so much.

**Comment by stenson** [comment, marked as answer]
Glad you're enjoying it! Great question, and yes one of the blind spots of the current documentation.

Two different tacks to take here — the easiest is probably the `AsciiTimeline` which is meant to be a kind of minimal AfterEffects-style timeline done entirely with text.

Here's a quick example:

```python
from coldtype import *
from coldtype.time.nle.ascii import AsciiTimeline

at = AsciiTimeline(4, """
                                <
[B    ]
            [A          ]
                [C         ]
                    [wght     ]
""")

@animation((1080, 540), timeline=at)
def ascii(f):
    def wdth(g):
        return at[g.c].io2(f.i, 8, "eeio")

    return (Glyphwise("ABC", lambda g:
        Style(Font.MutatorSans(), 250,
            wdth=wdth(g),
            wght=at["wght"].io2(f.i, 10, "eeio")))
        .align(f.a.r))
```

That produces this animation:

https://user-images.githubusercontent.com/89770/139151792-2169598a-3ee8-4040-829b-b31d0865b840.mp4

At first when I wrote the `AsciiTimeline` I thought it was kind of a "toy" solution to sequencing, but it turns out to be pretty useful in even longer-form stuff, like this timeline I worked up for a recent video loop (that hasn't been released yet).

![image](https://user-images.githubusercontent.com/89770/139150383-db0ec87f-19dc-4d7b-b222-5fbe545812cf.png)

There are a few examples in the coldtype example/animations folder as well:

- https://github.com/goodhertz/coldtype/blob/main/examples/animations/officehours.py
- https://github.com/goodhertz/coldtype/blob/main/examples/animations/ascii_choreography.py
- https://github.com/goodhertz/coldtype/blob/main/examples/animations/ascii_twostep.py
 
The other tack (better for longer-form animations) is to sequence them in something like Premiere or Blender (video editing mode). Both are supported but are a little heavier to set up completely, since they both involve another large program. Hoping to do some documentation of the Blender option in the not-too-distant future, but haven't had time to focus on it lately.

**Comment by Mark2Mark** [comment]
Oh dear, this looks gorgeous. I will have a play with the asciiTimeline today, that could be exactly it. Dang! Is there anything Coldtype can’t do?

I’ll probably stitch things together in DaVinci for the big picture, but first for the granular control of the single clips 😀

Thank you so much!

**Reply by stenson** [reply]
Cool! I've used DaVinci a little bit but not a lot — has a pretty nice Python API tho so I'm sure it wouldn't be too hard to get it plugged into the Coldtype sequencing code

**Comment by Mark2Mark** [comment]
BTW: I am also having quite some fun _decoding_ (in a literal sense) your code :D 

This works great!
I wonder: could it be, that if dealt with one letter in a timeline, it cannot be dealt with once more?
Examples I tried:
```py
at = AsciiTimeline(4, """
                                          <
[A              ]
                                  [A     ]
"""
```
```py
at = AsciiTimeline(4, """
                                          <
[A              ]                 [A     ]
"""
```
In both cases, the second A does’t have any effect.

**Reply by stenson** [reply]
Hmm, interesting question regarding the multiples! Hadn't thought about that, but does make a lot of sense. At the moment, the timeline is basically a `dict`, so it's a one-to-one key mapping, but should be pretty straightforward to get back the "current" keyed item. Will make an issue for that!

**Reply by Mark2Mark** [reply]
Thanks so much!
